src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    /// <summary>
+    /// A singleton that tracks the open IVsWindowFrames and can report which documents are visible or active in a given <see cref="Workspace"/>.
+    /// Can be accessed via the <see cref="IDocumentTrackingService"/> as a workspace service.
+    /// </summary>
+    [Export]
+    internal class VisualStudioActiveDocumentTracker : ForegroundThreadAffinitizedObject, IVsSelectionEvents, IDisposable
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+
+        /// <summary>
+        /// Collection of all asynchronous tasks that are started by this service. This should only be tasks that are implicitly
+        /// async as we fetch services from the <see cref="IAsyncServiceProvider"/>, and are used to ensure we wait for them during shutdown.
+        /// These should not be waited for in calls to <see cref="GetActiveDocument(Workspace)"/> or <see cref="GetVisibleDocuments(Workspace)"/>
+        /// because those are expected to not be jumping to the UI thread per traditional Roslyn semantics and may deadlock.
+        /// </summary>
+        private readonly JoinableTaskCollection _asyncTasks;
```

Comment:
> and are used to ensure we wait for them during shutdown.

Dumb question.  Why would we want to wait on anything during shitdown?

Reply:
This is considered "best practice" by the shell team, even though Roslyn often flaunts it. With the shell getting more async, we should follow along. There have been plenty of shutdown bugs across VS where somebody does a fire-and-forget, and then that code has ran _after_ the shell has already shut some things down. For example, it's possible to have .NET code running in VS _after_ COM has already been shutdown, so using a COM service might end poorly.

Reply:
Gotcha.  Thanks.  Makes sense in a very unfortunate way.